docs: add comments for `UpdateResult`

movahhedi opened the pull request:
Closes #988.
References #428, #431.

Based on:
https://github.com/kysely-org/kysely/blob/deea8e22aae3dc4e142f055b6ec4d9a50a402bda/src/driver/database-connection.ts#L23-L34

igalklebanov commented on src/query-builder/update-result.ts:
```diff
@@ -1,5 +1,14 @@
+
+  /**
+   * The number of rows the update query changed.
+   * This is **optional** and only provided by some drivers like `node-mysql2`.
```
```suggestion
   * This is **optional** and only supported in dialects such as MySQL.
```

movahhedi replied:
Would you suggest the same change for [`database-connection.ts`](https://github.com/kysely-org/kysely/blob/deea8e22aae3dc4e142f055b6ec4d9a50a402bda/src%2Fdriver%2Fdatabase-connection.ts#L23-L34)?

https://github.com/kysely-org/kysely/blob/deea8e22aae3dc4e142f055b6ec4d9a50a402bda/src%2Fdriver%2Fdatabase-connection.ts#L23-L34

igalklebanov replied:
Yes. changed rows also exists in `mysql`, not just `mysql2`, and its probably part of the MySQL protocol in general.

movahhedi replied:
Applied.

movahhedi commented:
Although I've written the comment, I haven't yet truly understood what "Changed Rows" means.
I've seen these words:
- Matched rows
- Affected rows
- Updated rows
- Changed rows

What do each mean? Can we make the comment any better?

igalklebanov commented:
> * **Matched rows**: Rows that match the condition[s] of any query.

no such concept.

> * **Affected rows**: Rows that have been affected (inserted, updated [even if not changed], or deleted) by an `INSERT`/`UPDATE`/`DELETE` query.

kind of. it varies per query kind.
https://dev.mysql.com/doc/c-api/8.0/en/mysql-affected-rows.html

> * **Updated rows**: Rows affected by an `UPDATE` query, even if the content of it hasn't changed after the execution of the query.

No such concept outside of kysely. It is the `affected rows` value under a different, friendlier, and fitting to context, name.

> * **Changed rows**: Rows affected by an `UPDATE` query that had their content changed.

I'm not sure this is only relevant to `UPDATE`. Might be also relevant to `INSERT INTO ... ON DUPLICATE KEY UPDATE`.

igalklebanov commented on src/query-builder/update-result.ts:
```diff
@@ -1,5 +1,14 @@
   readonly numUpdatedRows: bigint
+
+  /**
+   * The number of rows the update query changed.
+   * This is **optional** and only supported in dialects such as MySQL.
```
```suggestion
   * The number of rows the update query changed.
   *
   * This is **optional** and only supported in dialects such as MySQL.
```
